fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java

```diff
@@ -310,4 +316,41 @@ OptExprBuilder castOutputColumnsTypeToTargetColumns(ColumnRefFactory columnRefFa
         }
         return root.withNewRoot(new LogicalProjectOperator(new HashMap<>(columnRefMap)));
     }
+
+    /**
+     * OlapTableSink may be executed in multiply fragment instances of different machines
+     * For non-duplicate key types, we must guarantee that the orders of the same key are
+     * exactly the same. In order to achieve this goal, we can perform shuffle before TableSink
+     * so that the same key will be sent to the same fragment instance
+     */
+    private PhysicalPropertySet createPhysicalPropertySet(InsertStmt insertStmt,
+                                                          List<ColumnRefOperator> outputColumns) {
+        if (!(insertStmt.getTargetTable() instanceof OlapTable)) {
+            return new PhysicalPropertySet();
+        }
+
+        OlapTable table = (OlapTable) insertStmt.getTargetTable();
+
+        if (KeysType.DUP_KEYS.equals(table.getKeysType())) {
+            return new PhysicalPropertySet();
+        }
+
+        List<Column> columns = table.getFullSchema();
+        Preconditions.checkState(columns.size() == outputColumns.size(),
+                "outputColumn's size must equal with table's column size");
+
+        List<Column> keyColumns = table.getKeyColumnsByIndexId(table.getBaseIndexId());
+        List<Integer> keyColumnIds = Lists.newArrayList();
+        keyColumns.forEach(column -> {
+            int index = columns.indexOf(column);
+            Preconditions.checkState(index >= 0);
+            keyColumnIds.add(outputColumns.get(index).getId());
+        });
+
+        HashDistributionDesc desc =
+                new HashDistributionDesc(keyColumnIds, HashDistributionDesc.SourceType.SHUFFLE_JOIN);
```

Comment:
SHUFFLE_AGG is enough?

Reply:
yes, strict order is unnecessary, I'll update